### PR TITLE
refactor: added source locations for `Def`, `Sig` and `Op`

### PR DESCRIPTION
**Pull request by LoZander**
added source locations for Def, Sig and Op AST nodes where these are missing. These already do exist for some ASTs, but are removed during the Naming phase `Namer.scala`.

Related to the work being done in #8837

### CODE REVIEW

**Comment by LoZander** [comment]
While working on this, I have noticed multiple locations where the locations of the `sym` or `spec` of1 `Def`, `Sig` and `Op` are used where it seems like it ought to have been the location of the `Def`, `Sig` and `Op` themselves. Perhaps, these places should be investigated.

**Comment by magnus-madsen** [comment]
@mlutze Can you review?

File: main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala [hunk 1]
```diff
@@ -217,7 +217,7 @@ object SemanticTokensProvider {
     * Returns all semantic tokens in the given definition `defn0`.
     */
   private def visitDef(defn0: TypedAst.Def): Iterator[SemanticToken] = defn0 match {
-    case Def(sym, spec, exp) =>
+    case Def(sym, spec, exp, loc) =>
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:220]: unused?

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/HoleCompletion.scala [hunk 2]
```diff
@@ -55,7 +55,7 @@ object HoleCompletion {
     )
 
     val matches = root.defs.values.flatMap {
-      case TypedAst.Def(sym, spec, _) =>
+      case TypedAst.Def(sym, spec, _, loc) =>
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/HoleCompletion.scala:58]: unused?

**Review by mlutze — APPROVED** [review]

**Comment by mlutze** [comment]
We can remove the loc from `spec` as a follow-up.

**Comment by LoZander** [comment]
> We can remove the loc from `spec` as a follow-up.

I'm not sure I understand why we'd removed it from `spec`. For instance, shouldn't location for `Def` be different from that of it's `Spec`?

**Comment by LoZander** [comment]
> > We can remove the loc from `spec` as a follow-up.
> 
> I'm not sure I understand why we'd removed it from `spec`. For instance, shouldn't location for `Def` be different from that of it's `Spec`?

I was under the impression that the `Spec` was only the signature. So that a the location of a `spec` for a `def` would not contain the location for the `exp` for instance. But I see now that during naming, the `loc` of the `def` is actually just "moved" to the spec. So the `spec` of a `def` will always have the same location as that of the `def`. I assume the same is true for `Op` and `Sig`. And in such a case, I suppose there is no reason to have loc in both the `Def`/`Sig`/`Op` and its spec

**Comment by LoZander** [comment]
I guess my final question is whether or not the location for a `Def`/its `Spec` contains the location for the body exp?

**Comment by magnus-madsen** [comment]
@LoZander Will you fix Matt's comments and remove loc from spec?

**Comment by magnus-madsen** [comment]
> > We can remove the loc from `spec` as a follow-up.
> 
> I'm not sure I understand why we'd removed it from `spec`. For instance, shouldn't location for `Def` be different from that of it's `Spec`?

No. Its not. That's why we want to move it back to the Def/Sig/Op.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala [hunk 3]
```diff
@@ -230,7 +230,7 @@ object SemanticTokensProvider {
     * Returns all semantic tokens in the given signature `sig0`.
     */
   private def visitSig(sig0: TypedAst.Sig): Iterator[SemanticToken] = sig0 match {
-    case TypedAst.Sig(sym, spec, exp) =>
+    case TypedAst.Sig(sym, spec, exp, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:233]: This probably also unused.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:233]: @LoZander If you use intellij idea it should warn you about unused variables.

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:233]: > @LoZander If you use intellij idea it should warn you about unused variables.

But don't bother fixing all unused variables that you did not introduce yourself. That's out of scope for this PR.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:233]: Yes, just focus on your code.

**Comment by LoZander** [comment]
> @LoZander Will you fix Matt's comments and remove loc from spec?

Yep, sure. I was just waiting because I got the impression that the two of you were still mid-discussion on whether or not it should be done now or later :)

**Comment by LoZander** [comment]
> > > We can remove the loc from `spec` as a follow-up.
> > 
> > 
> > I'm not sure I understand why we'd removed it from `spec`. For instance, shouldn't location for `Def` be different from that of it's `Spec`?
> 
> No. Its not. That's why we want to move it back to the Def/Sig/Op.

yep. Realised a bit after posting that. My understanding of spec was a bit wrong before :)

**Comment by LoZander** [comment]
Is it a mistake that the SimplfiedAst created in `visitDef` in `Simplifier` is given the sourcelocation of the originals symbol (not spec or the def itself)?

File: main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala (hunk 3 above)
Comment (LoZander) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:233]: Yes, these ones slipped by me. I'll fix it :)

**Comment by LoZander** [comment]
> * [ ]  Remove loc from spec
> * [ ]  Remove unused local variables you have introduced

Both have been completed :)

**Comment by magnus-madsen** [comment]
Merged!

**Pull request merged**